Scalafmt 3.0.0 fails on Scala3 code starting with case statement

This template is a guideline, not a strict requirement.
- **Version**: 3.0.0
- **Integration**: scalafmt-native
- **Configuration**:
```
runner.dialect = scala3
version=3.0.0
maxColumn=120
project.git = true
```

## Steps

Given code like this:
```scala
  given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
    def lift = case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
  }
```

When I run scalafmt like this:

```bash
scalafmt
```

## Problem

Scalafmt errors out with the following

```
/home/mdedetrich/github/protoquill/quill-sql/src/main/scala/io/getquill/parser/Lifter.scala:110: error: illegal start of simple expression
    def lift = case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
```

## Expectation

I would like for scalafmt to not error out

## Workaround

None as far as I can tell, not even using `// format: off` and `// format: on` seems to work

## Notes

Original code where it fails https://github.com/getquill/protoquill/blob/master/quill-sql/src/main/scala/io/getquill/parser/Lifter.scala#L106-L109

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
@@ -2240,7 +2240,7 @@ class FormatOps(
           case t @ SplitAssignIntoParts((x: Term.PartialFunction, _)) =>
             Some(new OptionalBracesRegion {
               def owner = Some(t)
-              def splits = getSplitsForStats(ft, nft, x.cases, false)
+              def splits = getSplitsForStats(ft, nft, x.cases, nlOnly = true)
               def rightBrace = treeLast(x)
             })
           case _ => BlockImpl.create(ft, nft)
@@ -2436,15 +2436,16 @@ class FormatOps(
         nft: FormatToken,
         head: => Tree,
         tail: Seq[Tree],
-        allowMain: Boolean
+        allowMain: Boolean,
+        nlOnly: Boolean = false
     )(implicit
         fileLine: FileLine,
         style: ScalafmtConfig
     ): Option[Seq[Split]] =
       if (head.tokens.headOption.contains(nft.right)) Some {
-        val forceNL = shouldBreakInOptionalBraces(ft)
+        val forceNL = nlOnly || shouldBreakInOptionalBraces(ft)
         tail.lastOption.fold {
-          getSplits(ft, head, forceNL, useMain = allowMain)
+          getSplits(ft, head, forceNL, useMain = !nlOnly && allowMain)
         }(getSplits(ft, _, forceNL))
       }
       else None
@@ -2453,12 +2454,13 @@ class FormatOps(
         ft: FormatToken,
         nft: FormatToken,
         trees: Seq[Tree],
-        allowMain: Boolean = false
+        allowMain: Boolean = false,
+        nlOnly: Boolean = false
     )(implicit
         fileLine: FileLine,
         style: ScalafmtConfig
     ): Option[Seq[Split]] = trees.headOption.flatMap { head =>
-      getSplitsForStatsImpl(ft, nft, head, trees.tail, allowMain)
+      getSplitsForStatsImpl(ft, nft, head, trees.tail, allowMain, nlOnly)
     }
 
     private def getSplitsForIf(

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
@@ -2186,13 +2186,13 @@ class FormatOps(
           case t: Term.For =>
             Some(new OptionalBracesRegion {
               def owner = Some(t)
-              def splits = getSplitsForStats(ft, nft, t.enums, false)
+              def splits = getSplitsForStats(ft, nft, t.enums)
               def rightBrace = seqLast(t.enums)
             })
           case t: Term.ForYield =>
             Some(new OptionalBracesRegion {
               def owner = Some(t)
-              def splits = getSplitsForStats(ft, nft, t.enums, false)
+              def splits = getSplitsForStats(ft, nft, t.enums)
               def rightBrace = seqLast(t.enums)
             })
           case _ => BlockImpl.create(ft, nft)
@@ -2234,7 +2234,7 @@ class FormatOps(
           case t: Ctor.Secondary =>
             Some(new OptionalBracesRegion {
               def owner = Some(t)
-              def splits = getSplitsForStats(ft, nft, t.init, t.stats, true)
+              def splits = getSplitsForStatsImpl(ft, nft, t.init, t.stats, true)
               def rightBrace = if (t.stats.nonEmpty) treeLast(t) else None
             })
           case t @ SplitAssignIntoParts((x: Term.PartialFunction, _)) =>
@@ -2286,7 +2286,7 @@ class FormatOps(
           case t: Term.Try =>
             Some(new OptionalBracesRegion {
               def owner = Some(t)
-              def splits = getSplitsForStats(ft, nft, t.catchp, false)
+              def splits = getSplitsForStats(ft, nft, t.catchp)
               def rightBrace = seqLast(t.catchp)
             })
           case _ => None
@@ -2431,7 +2431,7 @@ class FormatOps(
       getSplits(ft, tree, forceNL, allowMain && !multiStat, danglingKeyword)
     }
 
-    private def getSplitsForStats(
+    private def getSplitsForStatsImpl(
         ft: FormatToken,
         nft: FormatToken,
         head: => Tree,
@@ -2441,25 +2441,25 @@ class FormatOps(
         fileLine: FileLine,
         style: ScalafmtConfig
     ): Option[Seq[Split]] =
-      if (head.tokens.headOption.contains(nft.right)) {
+      if (head.tokens.headOption.contains(nft.right)) Some {
         val forceNL = shouldBreakInOptionalBraces(ft)
-        tail.lastOption match {
-          case None => Some(getSplits(ft, head, forceNL, useMain = allowMain))
-          case Some(res) => Some(getSplits(ft, res, forceNL))
-        }
-      } else None
+        tail.lastOption.fold {
+          getSplits(ft, head, forceNL, useMain = allowMain)
+        }(getSplits(ft, _, forceNL))
+      }
+      else None
 
     private def getSplitsForStats(
         ft: FormatToken,
         nft: FormatToken,
         trees: Seq[Tree],
-        allowMain: Boolean
+        allowMain: Boolean = false
     )(implicit
         fileLine: FileLine,
         style: ScalafmtConfig
-    ): Option[Seq[Split]] =
-      if (trees.isEmpty) None
-      else getSplitsForStats(ft, nft, trees.head, trees.tail, allowMain)
+    ): Option[Seq[Split]] = trees.headOption.flatMap { head =>
+      getSplitsForStatsImpl(ft, nft, head, trees.tail, allowMain)
+    }
 
     private def getSplitsForIf(
         ft: FormatToken,

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
@@ -1890,7 +1890,7 @@ given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
     case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
 }
 >>>
-test does not parse
 given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
-  def lift = case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
-           ^
+  def lift =
+     case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+}

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
@@ -1882,3 +1882,15 @@ object a:
       def foo(s: String) =
         true
    end given
+<<< #2687 preserve break for partial function
+maxColumn = 100
+===
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift =
+    case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+}
+>>>
+test does not parse
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift = case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+           ^

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
@@ -1842,7 +1842,7 @@ given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
     case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
 }
 >>>
-test does not parse
 given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
-  def lift = case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
-           ^
+  def lift =
+     case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+}

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
@@ -1834,3 +1834,15 @@ object a:
    given Foo[String] with
       def foo(s: String) = true
    end given
+<<< #2687 preserve break for partial function
+maxColumn = 100
+===
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift =
+    case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+}
+>>>
+test does not parse
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift = case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+           ^

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
@@ -1922,3 +1922,15 @@ object a:
       def foo(s: String) =
         true
    end given
+<<< #2687 preserve break for partial function
+maxColumn = 100
+===
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift =
+    case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+}
+>>>
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift =
+     case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+}

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
@@ -2029,3 +2029,25 @@ object a:
    given Foo[String] with
       def foo(s: String) = true
    end given
+<<< #2687 preserve break for partial function
+maxColumn = 100
+===
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift =
+    case PropertyAlias(a, b) => '{ PropertyAlias(${ a.expr }, ${ b.expr }) }
+}
+>>>
+given liftPropertyAlias: NiceLiftable[PropertyAlias] with {
+  def lift =
+     case PropertyAlias(a, b) =>
+       '{
+         PropertyAlias(
+           ${
+             a.expr
+           },
+           ${
+             b.expr
+           }
+         )
+       }
+}

